crates/server/build.rs

```diff
@@ -1,12 +1,95 @@
-use std::process::Command;
+use std::path::{Path, PathBuf};
+use std::{env, fs};
+
+use cached_path::{Cache, Options};
+use zip::ZipArchive;
 
 fn main() {
-    let status = Command::new("sh")
-        .arg("../../scripts/download-node-ui.sh")
-        .status()
-        .expect("Failed to execute script");
+    let src = option_env!("CALIMERO_WEB_UI_SRC").unwrap_or(
+        "https://github.com/calimero-network/admin-dashboard/archive/refs/heads/master.zip",
+    );
+
+    let force = option_env!("CALIMERO_WEB_UI_FETCH")
+        .map(|v| v == "1" || v.eq_ignore_ascii_case("true"))
+        .unwrap_or(false);
+
+    let project_root = PathBuf::from(env::var("CARGO_MANIFEST_DIR").unwrap());
+    let static_files_target = project_root.join("../../node-ui/build");
+    let marker_file = project_root.join("build-deps-changed.txt");
+
+    let target_missing = !Path::new("target").exists();
+    let node_ui_missing = !static_files_target.exists();
+
+    if target_missing || node_ui_missing {
+        eprintln!("Triggering rebuild because `target/` or `node-ui/build` is missing.");
+        fs::write(&marker_file, format!("trigger: {:?}\n", std::time::SystemTime::now()))
+            .expect("Failed to write marker file");
+    }
+
+    let cache = Cache::builder()
+        .dir(PathBuf::from(
+            env::var("CARGO_TARGET_DIR").unwrap_or_else(|_| "target".into()),
+        ))
+        .build()
+        .expect("Failed to create cache");
+
+    let options = if force {
+        Options::default().extract()
+    } else {
+        Options::default()
+    };
```

Comment:
you already even use `extract` here, but I'm not sure I understand why we're only doing that when we're force-fetching